### PR TITLE
*: use `StaticExprContext` instead of mock context to build expression

### LINKED ISSUE #54271: Different partition locations for the same key when SQL mode is differnt
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
> set @@sql_mode='allow_invalid_dates';
> create table t(a datetime primary key) partition by range columns(a) (
    partition p0 values less than ('2020-02-31 00:00:00'),
    partition p1 values less than (MAXVALUE)
)
> insert into t values('2020-03-01 00:00:00')
> set @@sql_mode='';
> insert into t values('2020-03-01 00:00:00')
```

### 2. What did you expect to see? (Required)

The second insert should report an error because a is an primary key.

### 3. What did you see instead (Required)

The second insert succeeds. And we can see both of them can be found:

```
TiDB root@127.0.0.1:test> select * from t;
+---------------------+
| a                   |
+---------------------+
| 2020-03-01 00:00:00 |
| 2020-03-01 00:00:00 |
+---------------------+
2 rows in set
Time: 0.010s
TiDB root@127.0.0.1:test> select * from t partition(p0);
+---------------------+
| a                   |
+---------------------+
| 2020-03-01 00:00:00 |
+---------------------+
1 row in set
Time: 0.005s
TiDB root@127.0.0.1:test> select * from t partition(p1);
+---------------------+
| a                   |
+---------------------+
| 2020-03-01 00:00:00 |
+---------------------+
1 row in set
Time: 0.007s
```

The located in different partitions

### 4. What is your TiDB version? (Required)

Master. But release version may also have this issue.

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: ref #53388, close #54271

### What changed and how does it work?

Remove `mock.NewContext` and use `NewStaticExprContext` to build expression for partition table.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/table/tables/partition.go [hunk 1]
```diff
@@ -240,8 +240,8 @@ func initPartition(t *partitionedTable, def model.PartitionDefinition) (*partiti
 
 func newPartitionExpr(tblInfo *model.TableInfo, tp model.PartitionType, expr string, partCols []model.CIStr, defs []model.PartitionDefinition) (*PartitionExpr, error) {
 	// a partitioned table cannot rely on session context/sql modes, so use a default one!
-	ctx := mock.NewContext()
-	dbName := model.NewCIStr(ctx.GetSessionVars().CurrentDB)
+	ctx := contextstatic.NewStaticExprContext()
+	dbName := model.NewCIStr(ctx.GetEvalCtx().CurrentDB())
```
Comment (mjonss) [inline pkg/table/tables/partition.go:244]: Will this get the actual Current DB name? Or just the same default / empty string every time?

**Review by mjonss — APPROVED** [review]
LGTM, with one curious question :)

File: pkg/table/tables/partition.go (hunk 1 above)
Comment (lcwangchao) [reply pkg/table/tables/partition.go:244]: It is always an empty string and same with the previous behavior with `mock.Context`. (Seems it's useless, but just keep logic unchanged)

**Review by YangKeao — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

**Review by mjonss — APPROVED** [review]
LGTM, good find and explanation!

**Pull request merged**